### PR TITLE
add Parallel and Alternative constructors

### LINKED ISSUE #20: Use additional constructors for `StateMachine`
We would like to have more control over the structure of a state machine, being able not only to retrieve its state space.

Introducing new constructors like `Parallel :: StateMachine a b -> StateMachine c b -> StateMachine (a, c) (b, d)` and `Alternative :: StateMachine a b -> StateMachine c d -> StateMachine (Either a c) (Either b d)` would allow to keep more control over the structure of the state machine and determine its flow, while still preserving all the necessary information to compute the state space

### PR DESCRIPTION
**Pull request by marcosh**
closes #20

### CODE REVIEW

File: decision-log/2023-01-17-add-parallel-and-alternative-constructors.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  Mealy machine, at least in theory, admit an instance for `Strong` and
+  `Choice` type classes, to allow parallel and alternative composition.
+
+  In our current implementation, we could implement `Strong` and `Choice` just
+  by forwarding the composition to the `BaseMachine` layer.
+
+  Still everytime that we are combining two state machine, either sequentially,
+  in parallel or in alternative, the set of states is always the cartesian
+  product of the sets of states. It follows that we would not be able to
+  understand
```
Comment (AlexeiDrake) [inline decision-log/2023-01-17-add-parallel-and-alternative-constructors.yaml:13]: Unfinished sentence?

File: src/CRM/BaseMachine.hs [hunk 2]
```diff
@@ -107,16 +107,22 @@ instance Functor (ActionResult topology state initialVertex) where
   fmap f (ActionResult state output) =
     ActionResult state (f output)
 
--- ** Identity machine
+-- ** Stateless machines
 
--- | The `id` machine always outputs its input and never changes its state
-identity :: BaseMachine ('Topology '[ '( '(), '[ '()])]) a a
-identity =
+statelessBase :: (a -> b) -> BaseMachine TrivialTopology a b
```
Comment (AlexeiDrake) [inline src/CRM/BaseMachine.hs:112]:
```suggestion
-- | The `statelessBase` transforms its input to its output and never changes its state
statelessBase :: (a -> b) -> BaseMachine TrivialTopology a b
```

File: src/CRM/BaseMachine.hs [hunk 3]
```diff
@@ -107,16 +107,22 @@ instance Functor (ActionResult topology state initialVertex) where
...
   BaseMachine
     { initialState = InitialState STuple0
-    , action = ActionResult
+    , action = \state input ->
+        ActionResult state $ f input
     }
 
+-- ** Identity machine
+
+-- | The `id` machine always outputs its input and never changes its state
```
Comment (AlexeiDrake) [inline src/CRM/BaseMachine.hs:122]:
```suggestion
-- | The `identity` machine simply outputs its input and never changes its state. It is the result of `statelessBase id`.
```

File: src/CRM/StateMachine.hs [hunk 4]
```diff
@@ -25,6 +26,17 @@ data StateMachine input output where
     :: StateMachine a b
```
Comment (AlexeiDrake) [inline src/CRM/StateMachine.hs:26]: Would it make sense to call this constructor `Sequence` or  `Sequential`?

File: src/CRM/StateMachine.hs [hunk 5]
```diff
@@ -80,3 +92,13 @@ run (Compose machine1 machine2) a =
     (output2, machine2') = run machine2 output1
    in
     (output2, Compose machine1' machine2')
+run (Parallel machine1 machine2) a =
+  let
+    (output1, machine1') = run machine1 (fst a)
+    (output2, machine2') = run machine2 (snd a)
+   in
+    ((output1, output2), Parallel machine1' machine2')
```
Comment (AlexeiDrake) [inline src/CRM/StateMachine.hs:100]:
```suggestion
run (Parallel machine1 machine2) (a, b) =
  let
    (output1, machine1') = run machine1 a
    (output2, machine2') = run machine2 b
   in
    ((output1, output2), Parallel machine1' machine2')
```

File: src/CRM/Topology.hs [hunk 6]
```diff
@@ -59,3 +59,12 @@ instance {-# INCOHERENT #-} AllowedTransition ('Topology map) a b => AllowedTran
 
 instance {-# INCOHERENT #-} AllowedTransition topology a a where
   allowsTransition = AllowIdentityEdge
+
+-- ** Trivial topology
+
+$( singletons
```
Comment (AlexeiDrake) [inline src/CRM/Topology.hs:65]:
```suggestion
-- | The trivial topology only allows identity transitions.
$( singletons
```

File: src/CRM/StateMachine.hs [hunk 7]
```diff
@@ -41,32 +53,32 @@ instance Profunctor StateMachine where
...
   lmap f (Basic baseMachine) = Basic $ lmap f baseMachine
   lmap f (Compose machine1 machine2) = Compose (lmap f machine1) machine2
+  lmap f machine@(Parallel _ _) = Compose (stateless f) machine
+  lmap f machine@(Alternative _ _) = Compose (stateless f) machine
 
   rmap :: (b -> c) -> StateMachine a b -> StateMachine a c
   rmap f (Basic baseMachine) = Basic $ rmap f baseMachine
   rmap f (Compose machine1 machine2) = Compose machine1 (rmap f machine2)
+  rmap f machine@(Parallel _ _) = Compose machine (stateless f)
+  rmap f machine@(Alternative _ _) = Compose machine (stateless f)
```
Comment (AlexeiDrake) [inline src/CRM/StateMachine.hs:63]:
```suggestion
  lmap :: (a -> b) -> StateMachine b c -> StateMachine a c
  lmap f (Basic baseMachine) = Basic $ lmap f baseMachine
  lmap f (Compose machine1 machine2) = Compose (lmap f machine1) machine2
  lmap f machine = Compose (stateless f) machine


  rmap :: (b -> c) -> StateMachine a b -> StateMachine a c
  rmap f (Basic baseMachine) = Basic $ rmap f baseMachine
  rmap f (Compose machine1 machine2) = Compose machine1 (rmap f machine2)
  rmap f machine = Compose machine (stateless f)
```

File: src/CRM/StateMachine.hs (hunk 4 above)
Comment (marcosh) [reply src/CRM/StateMachine.hs:26]: Sometimes I think about it, too. I'm not sure, though. `Compose` recalls a lot the terminology used in category theory. `Sequential` gives a more graphical idea how what's happening.

I'd stick with `Compose` for now. I'll ask to some more people what they think

File: decision-log/2023-01-17-add-parallel-and-alternative-constructors.yaml (hunk 1 above)
Comment (marcosh) [reply decision-log/2023-01-17-add-parallel-and-alternative-constructors.yaml:13]: yes, looks like it :grin:

**Pull request merged**